Update Docs for installing a VM with Debian 11.02 and create setup-dev script (#1444)
votingworks/vxsuite: VirtualMachineSetup.md

# Virtual Machine Setup

> Follow the steps below to configure a Virtual Machine for development at VotingWorks. This guide assumes you are using Mac OS, if you are not modify the instructions as appropriate. 

Reviewer: lmk if you want me to write up how to do this on linux, in case someone doesn't want to use a mac.

Author: Cool, I think its fine for now but if someone mentions wanting to do it on linux then it would be great to have you help with that. I don't think we need to do it until there is a clear need though.

> Follow the steps below to configure a Virtual Machine for development at VotingWorks. This guide assumes you are using Mac OS, if you are not modify the instructions as appropriate. 

Reviewer 2: For consistency:
```suggestion
Follow the steps below to configure a Virtual Machine for development at VotingWorks. This guide assumes you are using macOS, if you are not modify the instructions as appropriate. 
```

## Create Debian Virtual Machine

1. Download and install [Parallels Pro](https://www.parallels.com/products/desktop/pro). For employees: You can expense this purchase. For small development tasks basic Parallels may be sufficient but we recommend Parallels Pro in order be able to increase the memory of the VM to higher levels, etc.

2. Download [Debian 11.2](https://www.debian.org/releases/bullseye/debian-installer/), bullseye. Download the `netinst CD image` appropriate for your processor. If your computer is has an ARM based processor, such as the Macbooks with M1 chips you **must** download the `arm64` image, otherwise you probably want the `amd64` image.

> 3. Create a new virtual machine in Parallels (File -> New). Chose to install an OS from a image file and navigate to the `.iso` image you downloaded in step 2. 

Reviewer 2: Unicode!
```suggestion
3. Create a new virtual machine in Parallels (File → New). Chose to install an OS from a image file and navigate to the `.iso` image you downloaded in step 2. 
```

 
 - **Hardware/Shared Printers:** Uncheck *Share Mac printers with Debian GNU/Linux*.

 - **Hardware/Sound & Camera:** Uncheck *Share Mac camera with Linux*

 - **Backup** Check the box for *SmartGuard* in order to enable automatic snapshots of your VM. This is highly recommended as long as you have the storage space. Note that you can also manually take snapshots at any time. Click *Details...* to customize how often snapshots are taken and how many are kept on your machine. If you aren't sure just keep the default settings.

Other settings that may not be present in newer versions of Parallels (in which case you can safely ignore)

- **Options/Applications:**  un-check Share Linux applications with Mac unless you enjoy clutter or have a specific need.
- **Options/More Options:** check Authenticate with macOS SSH public key.

When you are done, close the Configuration modal, and click *Continue*

## Installing Debian 

Once you boot your new virtual machine in Debian you will be prompted with the install guide. 

> 1. Chose either *Install* or *Graphical Install*, the latter will be used in this guide but it is just a slightly nicer UI, all of the steps and options are the same. 

Author: @umbernhard I wasn't sure if I should be doing Install with Speech Synthesis here, do you know if that mattes?

Reviewer: Install with speech synthesis just launches a text-to-speech engine for the installer, so it's only necessary if you need that during install time. If you install GNOME you still get speech synthesis for free regardless of how you install.

![Debian Install Screen](setup_screenshots/install_screen.png?raw=true)

2. Select your language, location, and keyboard layout on the next few screens. Debian will then begin installing. 

> 3. *Configure the network* You will be asked for a Hostname. Since we aren't running a network this doesn't really matter, you can leave the default `debian`.

Reviewer: This is another difference between prod and dev: on prod we need to think about what this name is, since LVM uses it for naming logical volumes. I've been using `VxMark`, tho I'm gonna switch it to `Vx` for use on all prod machines. Again, I'm interested to hear with @benadida and @eventualbuddha think about that. We could just use the default `debian` on prod, too, it's not ultimately _that_ important...

![Debian Install Screen](setup_screenshots/hostname.png?raw=true)

You will then be asked for a domain name, this can be left blank.

![Debian Install Screen](setup_screenshots/domain_name.png?raw=true)

4. *Set up users and passwords* When prompted chose a password for your root user. Debian will create both a `root` user and another user account for you. This security measure is a bit of overkill for us since we are just developing on this VM, so it may simplify things to use the same password for both. 

Next you will be prompted to set up your non-root user account. You will be asked for the real name of the user, a username, and a password. It will make things slightly simpler to use the same username as your home computers user account but you can chose anything you want here. 

Chose your timezone and click continue to continue the installation. 

> 6. *Partition Disks* You will be prompted on how to partition the disks. The production machines use LVM but for development it should be fine to just chose *Guided - use entire disk*. 

Author: @umbernhard Review my advice here.

Reviewer: that's right, I'd be interested to hear other's thoughts on whether we should use the same partitioning on dev as we do on prod. @benadida @eventualbuddha thoughts? It's not that much harder to set up.

Reviewer 3: I don't think this will matter all that much, especially since the file path layout *is* different in prof and dev and i don't anticipate they will become the same in dev and locked down... So in interest of simplicity I'd say dev could stick to just one partition and be done with it.

I'm open to counter arguments though!

Reviewer 2: I don't have a strong feeling on the matter.

![Debian Install - Disk Partitioning Screen](setup_screenshots/partitioning.png?raw=true)

Click continue on the next screen, confirming all data will be erased. 

When asked for what partitioning scheme to use select *All files in one partition*

![Debian Install - Disk Partitioning Screen](setup_screenshots/partitioning2.png?raw=true)

Select *Finish partitioning and write changes to disk* on the next confirmation screen, and then click 'Yes' when asked if you want to *Write the changes to disk?*

7. After the install you will be asked if you want to *Scan extra installation media?* You can select **No**. 

8. You will then be asked to configure the package manager with a Debian archive mirror. Assuming you are in the US, chose *United States* and then *deb.debian.org*. When asked for http proxy information you can leave it blank. 

> 9. You'll be asked whether or not to share statistics through the package usage survey, feel free to chose whatever you like. 

Reviewer: I don't think it really matters since our VMs are pretty disposable. I always say no.

![Software Install Screen](setup_screenshots/software.png?raw=true) 

11. When installation is complete you will be prompted to reboot your system. 

## Install Parallels Tools

Before you can continue with the Debian Quickstart in the main README we will want to install the Parallels tools, this will allow all of the Parallels features, such as copy and pasting between your VM and home computer to work. 

 Log into your new user account, with the user password you set. Click Actions -> Install Parallels Tools, and then click Continue on the pop-up. 

![Install Parallels Tools](setup_screenshots/install_parallels_tools.png?raw=true) 

 Open a terminal (Activities -> Terminal). And enter the command: 

`su - ` Note: the trailing `-` here is important!

Then authenticate with the *root* password. You are now in a root shell. The installation CD will auto-mount in a non-readable manner, to fix this enter the command:

```sh

> mount -t iso9660 -o exec /dev/cdrom /media/cdrom`

Reviewer 2: ```suggestion
mount -t iso9660 -o ro,exec /dev/cdrom /media/cdrom
```

```

> If you get an error about it already being mounted, first enter

Reviewer 2: ```suggestion
If you get an error about it already being mounted, first enter:
```

```sh
umount /media/cdrom
```

Then enter the following commands to start the installer: 

```sh
cd /media/cdrom
./install
``` 

Click continue through the prompts and your machine will automatically reboot at the end of this installation process. You can test that it works by copy and pasting things back and forth from the VM and your home machine. 

## Using the Terminal and VS Code on your home machine

> If you would be like to develop from your home machines terminal and use VS Code, you will need to do the following things. This assumes a Mac OS.

Reviewer 2: ```suggestion
If you would be like to develop from your home machine and use VS Code, you will need to do the following things. This assumes macOS.
```

On your home machine in the terminal open the ssh config file:
```sh
vim ~/.ssh/config
```
In your VM open the settings application. Click Network, then the gear icon next to the Wired Connected toggle. Copy the IPv4 address.

> In that file add a new entry like so, HOSTNAME will be how you get in and out of your VM, name it anything you would like.

Reviewer 2: ```suggestion
In that file add a new entry with a `<HOSTNAME>` placeholder for how address the VM (e.g. `ssh vx` if `<HOSTNAME>` is `vx`), name it anything you like (we recommend `vx`):
```

```
Host <HOSTNAME> 
  HostName <HOST_IPv4_ADDRESS>
  ForwardAgent yes
```

> If your username on the vm is different then your home computer you will also need to add a line

Reviewer 2: ```suggestion
If your username on the VM is different than your home computer you will also need to add a line:
```

```
  User <VM_USERNAME>
```

> Save the file and test the connection by typing the following command on your home machine

Reviewer 2: ```suggestion
Save the file and test the connection by typing the following command on your home machine:
```

```
ssh <HOSTNAME>
```

> 

Reviewer 2: ```suggestion

If the above command prompted you for a password, run this on your home machine:

\`\`\`
ssh-copy-id <HOSTNAME>
\`\`\`

Then try the `ssh` command again. It should complete without prompting for a password.
```

Author: Whoa, I've always just lived with putting my password in all the time lol

Download and install [VS Code](https://code.visualstudio.com/) if desired for development.
Install the [Remote SSH extension](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-ssh) within VS Code. You may need to restart VS Code.

Open *Remote Explorer* on the left sidebar. 

> Click the + button in the row with the VM hostname/alias to open a new remote window. Once connected, it should say "SSH: vx" in the bottom left of the window. It may take a moment to install tools the first time you do this. 

Reviewer 2: ```suggestion
Click the + button in the row with the VM hostname/alias to open a new remote window. Once connected, it should say "SSH: <HOSTNAME>" in the bottom left of the window. It may take a moment to install tools the first time you do this. 
```
